### DISCUSSION TITLE
Constant "SSL SYSCALL error: Connection timed out" when using a connection pool, application freezing for 15 minutes

### DISCUSSION THREAD

**Discussion by lmyslinski** [sqlalchemy/sqlalchemy#10903] [Usage Questions]
I'm trying to run a *very* basic FastAPI + SQLAlchemy application in production. Over the 2 past weeks I've read almost all of the documentation, reported outage issues with my compute provider and inspected every timeout setting available from my Postgres provider (Supabase). I'm out of options at this point as to what I can do make SQLAlchemy connection pooling be stable. 

To the point: My app receives requests & does some background processing, which can take roughly between 15-120 seconds. Because, as I've learned, the `sessionmaker` object is not thread safe, I'm not passing the `sessionmaker` dependency into my threaded work:

```
@router.get("/{uuid}")
def trigger_data_extraction(
...
    background_tasks: BackgroundTasks,
...
    db: Session = Depends(get_db),
):
    background_tasks.add_task(run_extraction...)
    set_file_status(uuid, "processing", db)
    return {"message": "Data extraction triggered"}
```

These instead, as I believe is done "by the book", only acquire a `scoped_session` object once they actually need to access the DB:

```
async def run_extraction(
    uuid: uuid.UUID,
):
    ...
    *Heavy processing logic A*
    ...
    with ThreadLocalSession() as db:
        ...
        set_file_status(uuid, "done", db)
      *Processing logic B*
        
        
    logger.info(f"Data extraction complete for uuid {uuid}")
 ```   

Here's how the DB is configred in both cases:

```
engine = create_engine(
    db_url,
    connect_args={"connect_timeout": 10},
    pool_timeout=5,
    pool_recycle=3600,
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
ThreadLocalSession = scoped_session(SessionLocal)

def get_db():
    db = SessionLocal()
    logger.info(f"Yielding DB connection")
    yield db

```

Every once in a while, at a totally random timing, the connection pool will **freeze my entire application for 15 minutes**, as it tries to (reconcile?) a connection. This is what it looks like from the log side:

Logs from the `run_extraction` function invocation:

- Log entry from the execution of logic A
- 2024-01-19 14:04:01 [debug    ] Connection <connection object at 0x7f35c405d800; dsn: 'user=postgres password=xxx connect_timeout=10 dbname=postgres host=db.tybgdmskgvlgxwjgbvsv.supabase.co port=5432', closed: 0> being returned to pool [sqlalchemy.pool.impl.QueuePool] filename=base.py func_name=_finalize_fairy module=base pathname=/app/.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py process=305 process_name=MainProcess request_id=None thread=139869074018304 thread_name=MainThread
- 2024-01-19 **14:04:01** [debug    ] Connection <connection object at 0x7f35c405d800; dsn: 'user=postgres password=xxx connect_timeout=10 dbname=postgres host=db.tybgdmskgvlgxwjgbvsv.supabase.co port=5432', closed: 0> rollback-on-return [sqlalchemy.pool.impl.QueuePool] filename=base.py func_name=_reset module=base pathname=/app/.venv/lib/python3.11/site-packages/sqlalchemy/pool/base.py process=305 process_name=MainProcess request_id=None thread=139869074018304 thread_name=MainThread

- 2024-01-19 **14:20:01** [info     ] Invalidate connection <connection object at 0x7f35c405d800; dsn: 'user=postgres password=xxx connect_timeout=10 dbname=postgres host=db.tybgdmskgvlgxwjgbvsv.supabase.co port=5432', closed: 2> (reason: OperationalError:could not receive data from server: Connection timed out 
- SSL SYSCALL error: Connection timed out
- Log with a different connection object being returned from pool & my processing completing successfully (albeit, 16 minutes later)

Honestly, I'm out of ideas on where to take this next other than to disable pooling and make a new connection every time I need the DB.

- I've tried using pre-ping, however that also results in 15 min freezes
- I've reduced the pool recycle time from 1h to 10 minutes in hope that the connection would always be fresh when pooled. Didn't help, as you can see here
- I've checked the connection pool size & limits on Supabase (I'm nowhere near the limits)
- I've parsed all of https://github.com/supabase/postgres to check whether they're adding any timeouts that I should take into account
- I didn't find any information about Postgres cutting long connections (as is the case for MySQL)
 
Any help is greatly appreciated.

**Comment by zzzeek** [comment]
do all problems go away when you use `NullPool` ?

**Reply by lmyslinski** [reply]
@zzzeek  Haven't tried that yet, but I believe so - the only problems I have are related to the way connection pooling is handled. I'm also connecting to the same DB from Next.js via Supabase's connection pooler, and I've had 0 issues with it.

**Reply by zzzeek** [reply]
how do you know if it's the connection pool if you haven't tried turning the pooling off?     or otherwise where is AsyncQueuePool hanging when it hangs since you know that's the problem ? 

also what's this?

    SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
    ThreadLocalSession = scoped_session(SessionLocal)

this is asyncio application?  why are you not using [AsyncSession](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#synopsis-orm) ? There's no "threads" in asyncio either, you can't use scoped_session() like that in asyncio.   This doesn't look at all like a SQLAlchemy asyncio application but the "def" itself is async, so this all looks very strange

**Reply by zzzeek** [reply]
as well as [create_async_engine](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.create_async_engine) - none of the non-asyncio API should be in a method that is an `async def` ....

**Reply by lmyslinski** [reply]
None of the DB code is async, so I've assumed that all the regular logic still stands. So just to recap, if I'm using both sync and async method, should I be using 2 separate engines?

**Reply by lmyslinski** [reply]
Sadly I haven't really found any examples on combing async processing with SQLAlchemy

**Reply by lmyslinski** [reply]
Honestly, it never dawned on my that I should be using async IO api

**Reply by lmyslinski** [reply]
Sounds like your reply at https://github.com/sqlalchemy/sqlalchemy/discussions/5923#discussioncomment-341870 would be quite relevant here

**Reply by lmyslinski** [reply]
Ok, this give me room to explore some options - thanks for the hints

**Reply by CaselIT** [reply]
> None of the DB code is async, so I've assumed that all the regular logic still stands. So just to recap, if I'm using both sync and async method, should I be using 2 separate engines?

your task is defined as `async def` so that is async code. If you use sync io there no other task will have a change of doing anything, since there's only a single thread when using asyncio.

>  if I'm using both sync and async method, should I be using 2 separate engines?

yes using two engines is the simplest option in that case

> Honestly, it never dawned on my that I should be using async IO api

well you are using async code, that usually requires async io

**Reply by adelhaddad** [reply]
@lmyslinski  Did you fix the issue ?

**Reply by lmyslinski** [reply]
Yes, since I've started using the async API's everything has been very stable:

```
async_engine = create_async_engine(db_url_async, pool_recycle=600)
async_session = async_sessionmaker(async_engine, expire_on_commit=False)

...
    async with async_session() as session:
        
        await session.execute(
...
        )
        await session.commit()
```
